### PR TITLE
Notify handle

### PR DESCRIPTION
**Pull request by sdaoden**
Funnily exactly 2 years after my private-mail attempt, now also on github (though i do not like it, just to lessen burden on people like you), here is the notify script patch again.
Note i have split the test out into a separate commit, i have not tried the test ever since.
The code i use ever since, too.

### CODE REVIEW

**Comment by sdaoden** [comment]
Merde. I try to look what has changed in the test series.

**Comment by sdaoden** [comment]
Ok now the tests pass with the new test framework.
I had forgotten the commented off-by-one (in _my_ code) that was still present in the original .. i fixed it and removed the comment.  Otherwise identical to the first series.

**Comment by troglobit** [comment]
Hi, sorry for the late reply on this.  I've been swamped with other work and also been changing $DAYJOB, so haven't had much time or energy to go through all PRs and issues on GitHub until now.  Also, as you mention, we have a bit of history from before and I've honestly been holding off reviewing this much because of that.

Now, I've checked out this your initial branch, looked over the code changes and test.  Besides from the actual review (coming up), I have the following general feedback:

  - Overall it looks good (!), even though I'm still not a fan of calling external scripts/programs from within syslogd.  I'm leaning towards guarding it with a configure option `--enable-notifiers`, which would also help keep down the size a bit.
  - I don't see you handling the `SIGHUP` case.  E.g., if a user removes a notifier from their .conf file, the old one still lingers, right?
  - You included a test, which is great, I wish all contributors did that.  However, the test needs to be redone to:
    - follow the structure of the other tests, i.e. use the `FAIL()`, `SKIP()`, etc. functions to ensure the correct return codes
    - not start a second syslogd but use the first one, this may need some refactoring, which I can help out with since I might that for another test/feature coming up

I'll do a detailed review now, using the GitHub PR review.

**Comment by sdaoden** [comment]
Hello, thanks for looking.
This is autotools right, i had to look, but surely could do this.
I think i adjusted the test from other tests around, that second instance i mean, iirc, but of course i can try to improve this if you desire!
Just keep on going.

File: src/syslogd.c [hunk 1]
```diff
@@ -3271,6 +3301,58 @@ static void logit(char *fmt, ...)
...
+
+		len = strlen(program);
+
+		np = calloc(1, sizeof(*np) + len +1);
+		if (np) {
+			/* xxx Actually wastes space -- vararray? */
+			np->n_program = (char*)&np[1];
+			memcpy(np->n_program, program, len);
+			SIMPLEQ_INSERT_TAIL(newn, np, n_link);
+		} else
```
Comment (troglobit) [inline src/syslogd.c:3322]: Please change this part to something as follows, handling errors first, and not trying to be clever:
```c
np = calloc(1, sizeof(*np));
if (!np) {
   ERR(...);
   return;
}
np->n_program = strdup(program);
if (!np->n_program) {
   free(np);
   ERR(...);
   return;
}
```
It's more readable and follows the general error handling style of syslogd.

File: src/syslogd.c [hunk 2]
```diff
@@ -3271,6 +3301,58 @@ static void logit(char *fmt, ...)
...
+
+	SIMPLEQ_FOREACH(np, &nothead, n_link) {
+		switch (fork()) {
+		case -1:
+			ERR("Cannot start notifier %s", np->n_program);
+			break;
+		case 0:
+			/* No specific I/O setup, just use what we had */
+			argv[0] = np->n_program;
+			argv[1] = (char*)logfile; /* logical unconst */
```
Comment (troglobit) [inline src/syslogd.c:3343]: No need to comment that.

File: src/syslogd.c [hunk 3]
```diff
@@ -3271,6 +3301,58 @@ static void logit(char *fmt, ...)
...
+
+static void notifier_invoke(const char *logfile)
+{
+	char *argv[3];
+	struct notifier *np;
+
+	logit("notify: rotated %s, invoking hooks\n", logfile);
+
+	SIMPLEQ_FOREACH(np, &nothead, n_link) {
+		switch (fork()) {
```
Comment (troglobit) [inline src/syslogd.c:3336]: Please save and, log PID, at least for debug purposes.

File: src/syslogd.c [hunk 4]
```diff
@@ -3271,6 +3301,58 @@ static void logit(char *fmt, ...)
...
+		case 0:
+			/* No specific I/O setup, just use what we had */
+			argv[0] = np->n_program;
+			argv[1] = (char*)logfile; /* logical unconst */
+			argv[2] = NULL;
+			execv(argv[0], argv);
+			_exit(1);
+		default:
+			/* Do not care beside that, no special process group
+			 * etc.; it will eventually be reaped via reapchild() */
```
Comment (troglobit) [inline src/syslogd.c:3349]: It's perfectly possible to reap children that are in their own process group.  I do that, e.g., in Finit.  So the comment is either redundant or incomplete.  I suggest dropping this comment too.

File: src/syslogd.c [hunk 5]
```diff
@@ -3271,6 +3301,58 @@ static void logit(char *fmt, ...)
...
+			logit("notify: forked for %s\n", np->n_program);
```
Comment (troglobit) [inline src/syslogd.c:3350]: Perhaps `logit("notify: starting %s[%d]\n", np->n_program, pid);` would be better for logging (see previous comment).

File: test/notify.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+cat <<EOF > ${CONF}
+notify ${NOT1}
+# Match all log messages, store in RC5424 format and rotate every 1 KiB
+*.*       -${LOG}    ;rotate=1k:2,RFC5424
+notify ${NOT2}
+EOF
+
+../src/syslogd -m1 -b :${PORT2} -d -sF -f ${CONF} -p ${SOCK2} -p ${ALTSOCK} -P ${PID2} >${LOG2} &
+sleep 3
+cat ${PID2} >> "$DIR/PIDs"
```
Comment (troglobit) [inline test/notify.sh:27]: Like I mentioned in the general comment, I'm planning a minor refactor of lib.sh so we can add generic arguments to the setup() function instead, meaning we don't need to start a second syslogd like you do here.  I'll see what I can do about that already later tonight.

File: test/notify.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	:
+else
+	exit $?
+fi
+
+if [ -x ../src/logger ]; then
+	:
+else
+	exit 0
+fi
```
Comment (troglobit) [inline test/notify.sh:39]: Please use FAIL() and SKIP() functions instead so the test harness is happy.  Also, the logger check should be moved to the top of the test, I guess (skipping if it's missing).

File: test/notify.sh [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+MSG=01234567890123456789012345678901234567890123456789
+MSG=$MSG$MSG$MSG$MSG$MSG$MSG$MSG$MSG$MSG$MSG
+../src/logger -u ${SOCK2} ${MSG}
+../src/logger -u ${SOCK2} 1${MSG}
+../src/logger -u ${SOCK2} 2${MSG}
+
+kill -9 `cat ${PID2}`
+
+sleep 1 # XXX synchronization of async process?
```
Comment (troglobit) [inline test/notify.sh:51]: I don't understand the comment, could you elaborate a bit, what did you see?

File: test/notify.sh [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+if [ -f ${LOG}.0 ] &&
+		grep 'script 1' ${NOT1STAMP} &&
+		grep 'script 2' ${NOT2STAMP}; then
+	:
+else
+	exit 1
+fi
```
Comment (troglobit) [inline test/notify.sh:58]: Please use FAIL() instead.

**Review by troglobit — CHANGES_REQUESTED** [review]

**Comment by sdaoden** [comment]
I will look into this next week, ok?
Thanks for doing this.

**Comment by troglobit** [comment]
Considering how long it took me to respond, of course! :)

That also gives me time to clean up the test code a bit.  Got quite a ways tonight, will continue tomorrow.

File: src/syslogd.c (hunk 1 above)
Comment (sdaoden) [reply src/syslogd.c]: Did that.

File: src/syslogd.c (hunk 2 above)
Comment (sdaoden) [reply src/syslogd.c]: Ok.

File: src/syslogd.c (hunk 3 above)
Comment (sdaoden) [reply src/syslogd.c]: ok, doing this via "childpid" as elsewhere in this file, and logging in same style, too.

File: src/syslogd.c (hunk 5 above)
Comment (sdaoden) [reply src/syslogd.c]: Hm, yes, as above.

File: test/notify.sh (hunk 6 above)
Comment (sdaoden) [reply test/notify.sh]: Did that. Much easier now to do this, thanks for your work last weekend!

File: test/notify.sh (hunk 7 above)
Comment (sdaoden) [reply test/notify.sh]: OK, FAIL and SKIP in use now.

File: test/notify.sh (hunk 8 above)
Comment (sdaoden) [reply test/notify.sh]: I removed the comment.  sleep is uses extensively to allow syslogd to do the write.  Well i do not know, if super-correct rocket-science-style testing shall be desired maybe a special test fifo that can be read from to sync maybe an idea?  sleep 3 works gracefully here.

File: test/notify.sh (hunk 9 above)
Comment (sdaoden) [reply test/notify.sh]: FAIL and OK, ok.

**Comment by sdaoden** [comment]
Thanks again for your interest in adding this.
On my github fork there is also the suggested notify_optional implementation in a branch of that name.  I .. would rather not open a pull request, i find it makes the code ugly due to all the conditionals?

File: test/notify.sh (hunk 8 above)
Comment (troglobit) [reply test/notify.sh]: True. Let's keep it simple for now.

**Comment by troglobit** [comment]
> On my github fork there is also the suggested notify_optional implementation in a branch of that name. I .. would rather not open a pull request, i find it makes the code ugly due to all the conditionals?

OK, I'll have a look later, but you''re probably right.

**Review by troglobit — APPROVED** [review]

**Pull request merged**